staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -682,28 +715,83 @@ func (s *Controller) nodeSyncLoop() {
 	// round.
 	s.servicesToUpdate = s.cache.allServices()
 	numServices := len(s.servicesToUpdate)
-	s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, newHosts)
+	s.servicesToUpdate = s.updateLoadBalancerHosts(s.servicesToUpdate, workers)
 	klog.V(2).Infof("Successfully updated %d out of %d load balancers to direct traffic to the updated set of nodes",
 		numServices-len(s.servicesToUpdate), numServices)
 
 	s.knownHosts = newHosts
 }
 
-// updateLoadBalancerHosts updates all existing load balancers so that
-// they will match the list of hosts provided.
-// Returns the list of services that couldn't be updated.
-func (s *Controller) updateLoadBalancerHosts(services []*v1.Service, hosts []*v1.Node) (servicesToRetry []*v1.Service) {
-	for _, service := range services {
+type ResultContext struct {
+	service *v1.Service
+	retry   bool
+}
+
+func (s *Controller) nodeSyncWorker(svcCh chan *v1.Service, resultCh chan ResultContext) {
+	klog.V(4).Infof("nodeSyncWorker Started")
+	for svc := range svcCh {
 		func() {
-			if service == nil {
+			result := ResultContext{service: svc}
+			defer func() {
+				resultCh <- result
+			}()
+			if svc == nil {
 				return
 			}
-			if err := s.lockedUpdateLoadBalancerHosts(service, hosts); err != nil {
-				runtime.HandleError(fmt.Errorf("failed to update load balancer hosts for service %s/%s: %v", service.Namespace, service.Name, err))
-				servicesToRetry = append(servicesToRetry, service)
+			klog.V(4).Infof("nodeSyncWorker processing service %s/%s", svc.Namespace, svc.Name)
+			hosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
+			if err != nil {
+				runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
+				result.retry = true
+				return
+			}
+
+			if err := s.lockedUpdateLoadBalancerHosts(svc, hosts); err != nil {
+				runtime.HandleError(fmt.Errorf("failed to update load balancer hosts for service %s/%s: %v", svc.Namespace, svc.Name, err))
+				result.retry = true
+				return
```

Comment:
You can also simplify this function if you use the approach I mentioned below.

Reply:
Done